### PR TITLE
drivers: can: socketCAN: move code from drivers to its own file

### LINKED ISSUE #42235: SocketCAN not supported for NUCLEO H743ZI
Running the sample **zephyr/samples/net/sockets/can/** with build and flash commands,

```
west build -p -b nucleo_h743zi zephyr/samples/net/sockets/can/
west flash
```

fails after flashing and the board outputs.

```
*** Booting Zephyr OS build v3.0.0-rc1-2-g140837ce3ac6  ***


[00:00:02.005,000] <err> net_socket_can_sample: No CANBUS network interface found!
[00:00:02.015,000] <err> net_socket_can_sample: Cannot start CAN application (-2)
uart:~$ 
```

I found a similar issue that was report back in 2020, https://github.com/zephyrproject-rtos/zephyr/issues/30432. Commits were added to fix this issue. 

Is there socket can support for the Nucleo H743zi?

### PR DESCRIPTION
**Pull request by alexanderwachter**
Cleanup the drivers from socketCAN and make its own driver that instantiates the net interface.

### CODE REVIEW

File: samples/net/sockets/can/Kconfig [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Private config options for can sample app
+
+# Copyright (c) 2018 Alexander Wachter
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "SocketCAN sample application"
+
+config LOOPBACK_MODE
```
Comment (jukkar) [inline samples/net/sockets/can/Kconfig:8]: For the sample specific Kconfig options in the networking samples, I added `SAMPLE` to the option name in order to make it obvious that this option is strictly sample specific. So I am suggesting to call this `CONFIG_SAMPLE_CAN_LOOPBACK_MODE` here?

File: subsys/net/lib/sockets/sockets_can.c [hunk 2]
```diff
@@ -95,20 +95,23 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
...
+	struct zcan_frame *zframe = (struct zcan_frame *)net_pkt_data(pkt);
+	struct can_frame frame;
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(receivers); i++) {
-		struct zcan_frame *zframe =
-			(struct zcan_frame *)net_pkt_data(pkt);
-		struct can_frame frame;
+	if (net_pkt_get_len(pkt) != sizeof(struct zcan_frame)) {
+		NET_ERR("pkt size missmatch");
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets_can.c]: Please avoid `NET_` prefixed macros in the samples, we have been trying to limit their use to core net code only. So better here to do `LOG_ERR`. The reason for this is that the sample code can be directly added to user application and user does not need to tweak the log entries or add any additional network specific config options.

File: subsys/net/lib/sockets/sockets_can.c [hunk 3]
```diff
@@ -327,7 +330,7 @@ static ssize_t zcan_recvfrom_ctx(struct net_context *ctx, void *buf,
 
 	if (net_pkt_read(pkt, (void *)&zframe, sizeof(zframe))) {
 		net_pkt_unref(pkt);
-
+		NET_DBG("cannot read pkt %p, len %zu", pkt, recv_len);
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets_can.c]: Same here, please use `LOG_DBG` instead

File: subsys/net/lib/sockets/sockets_can.c [hunk 4]
```diff
@@ -95,23 +95,20 @@ static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
 	 * which will give the packet to correct net_context(s).
```
Comment (jukkar) [inline subsys/net/lib/sockets/sockets_can.c:95]: The commit message `samples: net: sockets: socketCAN: add loopback mode` of the 3rd commit does not seem to be correct here. It talks about sample, but this change is for socket CAN core code.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/lib/sockets/sockets_can.c (hunk 2 above)
Comment (alexanderwachter) [reply subsys/net/lib/sockets/sockets_can.c]: This is net code. It is in the sockets where everything is logged with NET_

File: subsys/net/lib/sockets/sockets_can.c (hunk 4 above)
Comment (alexanderwachter) [reply subsys/net/lib/sockets/sockets_can.c:95]: Sorry, this is not supposed to go in.

File: samples/net/sockets/can/Kconfig (hunk 1 above)
Comment (alexanderwachter) [reply samples/net/sockets/can/Kconfig]: Done

**Comment by alexanderwachter** [comment]
Doc CI fail is unrelated.

File: drivers/can/can_socketcan.c [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+static void socket_can_recv(struct zcan_frame *frame, void *arg)
+{
+	struct socket_can_context *ctx = (struct socket_can_context *)arg;
+	struct net_pkt *pkt;
+	int ret;
+
+	LOG_DBG("pkt on interface %p", ctx->iface);
+	pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, sizeof(struct zcan_frame),
+					   AF_CAN, 0, K_NO_WAIT);
+	if (!pkt) {
```
Comment (henrikbrixandersen) [inline drivers/can/can_socketcan.c:37]:
```suggestion
	if (pkt == NULL) {
```

File: drivers/can/can_socketcan.c [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+	socket_context->iface = iface;
+
+	LOG_DBG("Init CAN interface %p dev %p", iface, dev);
+}
+
+static int socket_can_init(const struct device *dev)
+{
+	const struct socket_can_config *cfg = dev->config;
+
+	ARG_UNUSED(cfg);
```
Comment (henrikbrixandersen) [inline drivers/can/can_socketcan.c:136]: `cfg` is actually used in the `LOG_DBG()` macro call.

File: drivers/can/can_socketcan.c [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+static struct socket_can_context socket_can_ctx;
+
+static const struct socket_can_config socket_can_cfg = {
+	.can_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_canbus))
+};
+
+NET_DEVICE_INIT(socket_can, "SOCKET_CAN", socket_can_init,
+		NULL, &socket_can_ctx, &socket_can_cfg,
+		CONFIG_CAN_INIT_PRIORITY,
```
Comment (henrikbrixandersen) [inline drivers/can/can_socketcan.c:159]: Won't having this at the same initialisation priority as the CAN controller drivers cause a race condition?

File: samples/net/sockets/can/Kconfig [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# Private config options for can sample app
+
+# Copyright (c) 2018 Alexander Wachter
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "SocketCAN sample application"
+
+config SAMPLE_SOCKET_CAN_LOOPBACK_MODE
+	bool "Loopback LOOPBACK_MODE"
```
Comment (henrikbrixandersen) [inline samples/net/sockets/can/Kconfig:9]:
```suggestion
	bool "Loopback mode"
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Nice clean-up, just a few minor observations.

File: drivers/can/can_socketcan.c (hunk 6 above)
Comment (alexanderwachter) [reply drivers/can/can_socketcan.c:136]: But only if DBG is built.

File: samples/net/sockets/can/Kconfig (hunk 8 above)
Comment (alexanderwachter) [reply samples/net/sockets/can/Kconfig]: seems someone did find/replace on "mode" *wasn't me*

File: drivers/can/can_socketcan.c (hunk 7 above)
Comment (alexanderwachter) [reply drivers/can/can_socketcan.c]: Good point. Was always like this.
Reused the duplicated init priority instead

File: drivers/can/Kconfig [hunk 9]
```diff
@@ -60,14 +60,15 @@ config CANFD_MAX_DLC
...
+	int "socketCAN net device init priority"
+	default 81
+	depends on NET_SOCKETS_CAN
 	help
-	  CAN device driver initialization priority.
+	  socketCAN net device initialization priority.
 	  Do not mess with it unless you know what you are doing.
 	  Note that the priority needs to be lower than the net stack
-	  so that it can start before the networking sub-system.
+	  and CAN driver so that it can start before the networking sub-system.
```
Comment (henrikbrixandersen) [inline drivers/can/Kconfig:71]: It needs to initialise _after_ the CAN driver, but _before_ the network subsystem, right?

File: drivers/can/can_socketcan.c (hunk 6 above)
Comment (henrikbrixandersen) [reply drivers/can/can_socketcan.c:136]: Still used, though.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Comment by henrikbrixandersen** [comment]
Please rebase to get the documentation build fix from #41865.

**Review by jukkar — DISMISSED** [review]

File: drivers/can/Kconfig (hunk 9 above)
Comment (alexanderwachter) [reply drivers/can/Kconfig]: Yep, think so

File: samples/net/sockets/can/src/main.c [hunk 10]
```diff
@@ -12,8 +12,12 @@ LOG_MODULE_REGISTER(net_socket_can_sample, LOG_LEVEL_DBG);
 #include <net/socket.h>
 #include <net/socket_can.h>
 
+#ifdef SAMPLE_SOCKET_CAN_LOOPBACK_MODE
+#include <drivers/can.h>
+#endif
+
 #define PRIORITY  k_thread_priority_get(k_current_get())
-#define STACKSIZE 1024
+#define STACKSIZE 2048
```
Comment (dleach02) [inline samples/net/sockets/can/src/main.c:20]: just curious but was there a particular driver that caused this stack to have to grow or is this a general result of the split out?

Comment (alexanderwachter) [reply samples/net/sockets/can/src/main.c:20]: I had a kernel panic when I tested on HW. Not sure anymore if it was the stack.

File: drivers/can/Kconfig [hunk 11]
```diff
@@ -60,14 +60,17 @@ config CANFD_MAX_DLC
...
+	default 81
+	depends on NET_SOCKETS_CAN
 	help
-	  CAN device driver initialization priority.
+	  socketCAN net device initialization priority.
 	  Do not mess with it unless you know what you are doing.
 	  Note that the priority needs to be lower than the net stack
-	  so that it can start before the networking sub-system.
+	  so that it can start before the networking sub-system,
+	  and higher than the CAN driver sot that it can use the underlaying
```
Comment (henrikbrixandersen) [inline drivers/can/Kconfig:72]:
```suggestion
	  and higher than the CAN driver so that it can use the underlaying
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
One minor nit, otherwise looks good.

**Comment by alexanderwachter** [comment]
Just tested on nucleo_g474 and could not get it working.
Need to investigate what changed during the submission of this patch and now.

**Comment by henrikbrixandersen** [comment]
> Just tested on nucleo_g474 and could not get it working. Need to investigate what changed during the submission of this patch and now.

What is the nature of the errors, you are seeing? I have just tested this PR with `samples/net/sockets/can` on two `twr_ke18f` boards without issues.

**Comment by alexanderwachter** [comment]
Tx issues.
API sample works fine.

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
> Tx issues. API sample works fine.

@alexanderwachter Any progress on this? Any more information on the errors, you are seeing? Are they also present on `main`?

**Comment by henrikbrixandersen** [comment]
> > Tx issues. API sample works fine.
> 
> @alexanderwachter Any progress on this? Any more information on the errors, you are seeing? Are they also present on `main`?

@alexanderwachter Please provide further details on the issues you are seeing?

**Comment by alexanderwachter** [comment]
@henrikbrixandersen it was a stack overflow, caused by log immediate and a -115. I used the wrong symbol for the loopback Kconfig (forgot the CONFIG_ prefix).
However, works now!

**Review by jukkar — DISMISSED** [review]

File: drivers/can/CMakeLists.txt [hunk 12]
```diff
@@ -18,3 +18,4 @@ zephyr_library_sources_ifdef(CONFIG_USERSPACE        can_handlers.c)
 zephyr_library_sources_ifdef(CONFIG_CAN_SHELL        can_shell.c)
 
 add_subdirectory(transceiver)
+zephyr_library_sources_ifdef(CONFIG_NET_SOCKETS_CAN  can_socketcan.c)
```
Comment (henrikbrixandersen) [inline drivers/can/CMakeLists.txt:21]:
```suggestion
zephyr_library_sources_ifdef(CONFIG_CAN_SHELL        can_shell.c)
zephyr_library_sources_ifdef(CONFIG_NET_SOCKETS_CAN  can_socketcan.c)

add_subdirectory(transceiver)
```

File: drivers/can/Kconfig [hunk 13]
```diff
@@ -66,14 +66,17 @@ config CANFD_MAX_DLC
 	  optimize memory consumption.
 endif # CAN_FD_MODE
 
-config CAN_INIT_PRIORITY
-	int "CAN driver init priority"
-	default 80
+config CAN_SOCKETCAN_NETDEVICE_INIT_PRIORITY
```
Comment (henrikbrixandersen) [inline drivers/can/Kconfig:69]: Why not just `CAN_SOCKETCAN_INIT_PRIORITY`?

File: samples/net/sockets/can/Kconfig [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# Private config options for can sample app
+
+# Copyright (c) 2018 Alexander Wachter
+# SPDX-License-Identifier: Apache-2.0
+
+mainmenu "SocketCAN sample application"
+
+config SAMPLE_SOCKET_CAN_LOOPBACK_MODE
```
Comment (henrikbrixandersen) [inline samples/net/sockets/can/Kconfig:8]:
```suggestion
config SAMPLE_SOCKETCAN_LOOPBACK_MODE
```

File: samples/net/sockets/can/src/main.c [hunk 15]
```diff
@@ -176,6 +180,10 @@ static int setup_socket(void)
 		return -ENOENT;
 	}
 
+#ifdef CONFIG_SAMPLE_SOCKET_CAN_LOOPBACK_MODE
+	can_set_mode(DEVICE_DT_GET(DT_CHOSEN(zephyr_canbus)), CAN_LOOPBACK_MODE);
```
Comment (henrikbrixandersen) [inline samples/net/sockets/can/src/main.c:184]: With loopback mode, we should be able to remove the `harness` from `samples/net/sockets/can/sample.yaml` and add a console fixture for validating the sample in CI.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**